lib/srv/db/common/databaseobjectimportrule/import_rules.go

```diff
@@ -0,0 +1,279 @@
+// Teleport
+// Copyright (C) 2023  Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package databaseobjectimportrule
+
+import (
+	"regexp"
+	"sort"
+	"strings"
+	"unicode"
+
+	"github.com/gravitational/trace"
+
+	dbobjectv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1"
+	dbobjectimportrulev1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobjectimportrule/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/label"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/services"
+	libutils "github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/typical"
+)
+
+// ApplyDatabaseObjectImportRules applies the given set of rules onto a set of objects coming from a same database.
+// Returns a fresh copy of a subset of supplied objects, filtered and modified.
+// For the object to be returned, it must match at least one rule.
+// The modification consists of application of extra labels, per matching mappings.
+// If there are any errors due to invalid label template, the corresponding objects will be dropped.
+func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObjectImportRule, database types.Database, objs []*dbobjectv1.DatabaseObject) ([]*dbobjectv1.DatabaseObject, map[string]error) {
```

Comment:
There can be as many errors here as the number of objects - hundreds or more. I'm worried that aggregating this will be messy, so I wanted to report each one separately. I agree that this isn't the easiest API to follow; what would you say to pass logger as an argument to ApplyDatabaseObjectImportRules instead? This way we can call it as we go, without having to return the map.

Reply:
> I agree that this isn't the easiest API to follow; what would you say to pass logger as an argument to ApplyDatabaseObjectImportRules instead?

Sure. Alternatively, change the function to do one object at a time. I think passing the logger is easier.

Reply:
Done.